File: web/js/live_feed.js
/**
 * Live feed — one place every view learns that something changed.
 *
 * Desktop: the Python backend (backend/live_feed.py) holds a single SSE
 * connection to the API and pushes each frame in through `receive_live_event`.
 * Web build: the browser opens the same stream itself with EventSource.
 *
 * Views fetch once when they mount and then subscribe, instead of re-polling on
 * a timer. `LiveFeed.on(type, handler)` returns an unsubscribe function, `'*'`
 * subscribes to every type, and `LiveFeed.last(type)` is the most recent payload
 * — the server opens each connection with a snapshot of current state, so a view
 * that mounts late still has something to read without waiting for a push.
 */
(function () {
    const STREAM_URL = 'https://api.aallyn.net/v1/events/stream';

    // Every type the stream emits. Named explicitly because EventSource
    // dispatches by event name — an unlisted type would arrive nowhere.
    const EVENT_TYPES = [
        'challenge', 'chaos', 'corruxion', 'fluxion', 'longshade', 'wild_mana',
        'stampy', 'daily_bonuses', 'activity', 'server_status', 'trove_news',
        'giveaways', 'game_update'
    ];

    const handlers = new Map();     // type -> Set<handler>
    const lastByType = new Map();

    const state = { connected: false };

    const emit = (type, data) => {
        for (const key of [type, '*']) {
            const set = handlers.get(key);
            if (!set) continue;
            for (const handler of Array.from(set)) {
                try {
                    handler(data, type);
                } catch (err) {
                    console.error(`[LiveFeed] handler for "${type}" failed`, err);
                }
            }
        }
    };

    const setConnected = (connected) => {
        if (state.connected === connected) return;
        state.connected = connected;
        document.dispatchEvent(new CustomEvent('live_feed_status', { detail: { connected } }));
    };

    const ingest = (type, data) => {
        if (type === '_status') {
            setConnected(!!(data && data.connected));
            return;
        }
        if (!data || typeof data !== 'object') return;
        lastByType.set(type, data);
        emit(type, data);
    };

    window.LiveFeed = {
        get connected() { return state.connected; },
        last(type) { return type ? lastByType.get(type) || null : Object.fromEntries(lastByType); },
        on(type, handler) {
            if (typeof handler !== 'function') return () => {};
            if (!handlers.has(type)) handlers.set(type, new Set());
            handlers.get(type).add(handler);
            return () => {
                const set = handlers.get(type);
                if (set) set.delete(handler);
            };
        }
    };

    // --- desktop: frames arrive from the Python backend ---------------------
    const hasEelBridge = !!window.eel && typeof window.eel.expose === 'function' && !window.BTT_WEB_MODE;
    if (hasEelBridge) {
        window.receive_live_event = (type, data) => ingest(type, data);
        eel.expose(window.receive_live_event, 'receive_live_event');

        // A page reload drops everything this module knew while the backend's
        // connection kept running, so seed from its snapshot rather than waiting
        // for the next push (chaos moves weekly — that could be days).
        document.addEventListener('DOMContentLoaded', async () => {
            try {
                const response = await eel.get_live_feed_status()();
                const data = (response && (response.data || response)) || {};
                setConnected(!!data.connected);
                Object.entries(data.events || {}).forEach(([type, payload]) => ingest(type, payload));
            } catch (err) {
                /* backend not up yet — the first real push seeds us instead */
            }
        });
        return;
    }

    // --- web build: connect straight to the stream --------------------------
    // EventSource reconnects on its own; nothing to schedule here.
    if (typeof EventSource !== 'function') return;
    let source = null;
    const connect = () => {
        source = new EventSource(STREAM_URL);
        source.onopen = () => setConnected(true);
        source.onerror = () => setConnected(false);
        EVENT_TYPES.forEach((type) => {
            source.addEventListener(type, (event) => {
                try {
                    const payload = JSON.parse(event.data);
                    ingest(type, payload && payload.data);
                } catch (err) {
                    /* malformed frame — skip it */
                }
            });
        });
    };
    document.addEventListener('DOMContentLoaded', connect);
})();

File: backend/live_feed.py
"""One Server-Sent Events connection to the TroveAPI, fanned out in-process.

The app already consumed this stream for desktop notifications; it now opens at
launch and feeds the UI as well. Views fetch once when they mount and then let
the stream tell them what actually changed, instead of re-polling on a timer.

Design:
  * A single background thread holds one long-lived streaming GET and parses the
    ``event:``/``data:`` frames. Running in Python (not a WebView timer) keeps it
    alive while the window is hidden in the tray, and costs one socket rather
    than one request per feed per tick.
  * Subscribers are plain callables ``(event_type, data) -> None`` invoked on the
    stream thread. Each is isolated: one raising doesn't starve the others.
  * The server opens every connection with a SNAPSHOT of current state, so a
    subscriber that attaches late -- or one that reconnects -- gets the whole
    picture within a few seconds without asking for it. Consumers that must not
    act twice on the same occurrence dedupe themselves (see
    ``event_notifications``); consumers that just mirror state don't care.
"""
from __future__ import annotations

import json
import threading
import time

import eel
import requests

from backend.home import KIWI_API_BASE
from backend.response import resp, standardize_response

STREAM_URL = f"{KIWI_API_BASE}/events/stream"

# Reconnect backoff after the stream drops.
_RECONNECT_SECONDS = 5.0
# Read timeout: the server sends a ``: ping`` keep-alive ~every 20s, so a longer
# gap means the socket is dead and we should reconnect.
_READ_TIMEOUT = 45.0


class LiveFeed:
    def __init__(self):
        self._subscribers = []
        self._lock = threading.Lock()
        self._thread = None
        self._connected = False
        self._last = {}          # event_type -> most recent data payload

    # --- public API -----------------------------------------------------
    def subscribe(self, callback):
        """Register ``callback(event_type, data)``. Safe to call before start."""
        with self._lock:
            self._subscribers.append(callback)

    def start(self):
        if self._thread and self._thread.is_alive():
            return
        self._thread = threading.Thread(target=self._run, name="live-feed", daemon=True)
        self._thread.start()

    def is_connected(self):
        with self._lock:
            return self._connected

    def snapshot(self, event_type=None):
        """Last payload seen per event type -- lets a consumer that starts late
        read current state without waiting for the next push."""
        with self._lock:
            if event_type is not None:
                return self._last.get(event_type)
            return dict(self._last)

    # --- stream thread --------------------------------------------------
    def _run(self):
        while True:
            try:
                self._connect_once()
            except Exception:
                pass
            self._set_connected(False)
            time.sleep(_RECONNECT_SECONDS)

    def _connect_once(self):
        headers = {"Accept": "text/event-stream", "User-Agent": "BetterTroveTools"}
        with requests.get(STREAM_URL, stream=True, timeout=(10, _READ_TIMEOUT), headers=headers) as response:
            response.raise_for_status()
            self._set_connected(True)
            event_type = None
            data_lines = []
            for raw in response.iter_lines(decode_unicode=True):
                if raw is None:
                    continue
                line = raw.rstrip("\r")
                if line == "":
                    if event_type and data_lines:
                        self._dispatch(event_type, "\n".join(data_lines))
                    event_type, data_lines = None, []
                    continue
                if line.startswith(":"):
                    continue  # keep-alive comment
                if line.startswith("event:"):
                    event_type = line[6:].strip()
                elif line.startswith("data:"):
                    data_lines.append(line[5:].lstrip())
                # ignore other fields (retry:, id:)

    def _set_connected(self, connected):
        with self._lock:
            if self._connected == connected:
                return
            self._connected = connected
            subscribers = list(self._subscribers)
        for callback in subscribers:
            try:
                callback("_status", {"connected": connected})
            except Exception:
                pass

    def _dispatch(self, event_type, data_str):
        try:
            payload = json.loads(data_str)
        except (ValueError, TypeError):
            return
        data = payload.get("data") if isinstance(payload, dict) else None
        if not isinstance(data, dict):
            return
        with self._lock:
            self._last[event_type] = data
            subscribers = list(self._subscribers)
        for callback in subscribers:
            try:
                callback(event_type, data)
            except Exception:
                pass


feed = LiveFeed()


def _forward_to_ui(event_type, data):
    """Hand every frame to the frontend. Fire-and-forget: with no window open
    (or none yet) eel has nowhere to send it, which is fine -- the next
    connection gets a fresh snapshot."""
    try:
        eel.receive_live_event(event_type, data)
    except Exception:
        pass


feed.subscribe(_forward_to_ui)


@eel.expose
@standardize_response
def get_live_feed_status():
    """Whether the live stream is currently up, plus the last payload seen for
    each event type. The frontend uses this on view mount: connected means it can
    stop polling, and the snapshot spares it a round of fetches."""
    connected = feed.is_connected()
    data = {"connected": connected, "events": feed.snapshot()}
    return resp(True, data=data, **data)
